src/Json/Decode/Extra.elm

```diff
@@ -285,3 +288,41 @@ lazy getDecoder =
   customDecoder value
     <| \rawValue ->
         decodeValue (getDecoder ()) rawValue
+
+
+{-| This function turns a list of decoders into a decoder that returns a list.
+
+The returned decoder will zip the list of decoders with a list of values, matching each decoder with exactly one value at the same position. This is most often useful in cases where `Json.Decode.oneOf`, which will try every decoder for every value in the list, would be too lenient.
+
+Note that this function, unlike `List.map2`'s behaviour, expects the list of decoders to have the same length as the list of values in the JSON.
+
+    type FloatOrInt
+        = I Int
+        | F Float
+
+    -- we'd like a list like [I, F, I] from this
+    -- fairly contrived example, but data like this does exist!
+    json = "[1, 2.0, 3]"
+
+    intDecoder = Decode.map I Decode.int
+    floatDecoder = Decode.map F Decode.float
+
+    decoder : Decoder (List FloatOrInt)
+    decoder =
+        sequence [ intDecoder, floatDecoder, intDecoder ]
+
+    decoded = Decode.decodeString decoder json
+    -- Ok ([I 1,F 2,I 3]) : Result String (List FloatOrInt)
+
+-}
+sequence : List (Decoder a) -> Decoder (List a)
+sequence decoders =
+  customDecoder
+    (list value)
+    (\jsonValues ->
+       if List.length jsonValues /= List.length decoders then
```

Comment:
This might better be expressed with a custom `List.map2`-like function to avoid walking 3 times n (that's the "worst case" by performance, but the case that means decoding success). But there's a reason that that function is implemented in native code in core – you can't really make it tail-recursive (at least I couldn't figure it out). Don't want people to make deep stacks or even blow the stack just by decoding a list of values...